crates/uv-python/fetch-download-metadata.py

```diff
@@ -540,6 +541,109 @@ async def _fetch_checksums(self, downloads: list[PythonDownload]) -> None:
+    PLATFORM_RE = re.compile(r"(\w+)-(\w+)\.(?:zip|tar\.gz)$", re.ASCII)
+
+    def __init__(self, client: httpx.AsyncClient):
+        self.client = client
+
+    async def find(self) -> list[PythonDownload]:
+        downloads = await self._fetch_downloads()
+        await self._fetch_checksums(downloads, n=10)
+        return downloads
+
+    async def _fetch_downloads(self) -> list[PythonDownload]:
+        resp = await self.client.get(self.RELEASE_URL)
+        resp.raise_for_status()
+        releases = resp.json()
+
+        results = {}
+        for release in releases:
+            m = self.GRAALPY_VERSION_RE.search(release["tag_name"])
+            if not m:
+                continue
+            graalpy_version = m.group(1)
+            m = self.CPY_VERSION_RE.search(release["body"])
+            if not m:
+                continue
+            python_version_str = m.group(1)
+            if not m.group(2):
+                python_version_str += ".0"
+            python_version = Version.from_str(python_version_str)
+            for asset in release["assets"]:
+                url = asset["browser_download_url"]
+                m = self.PLATFORM_RE.search(url)
+                if not m:
+                    continue
+                platform = self._normalize_os(m.group(1))
+                arch = self._normalize_arch(m.group(2))
+                libc = "gnu" if platform == "linux" else "none"
+                download = PythonDownload(
+                    release=0,
+                    version=python_version,
+                    triple=PlatformTriple(
+                        platform=platform,
+                            arch=arch,
+                            libc=libc,
+                        ),
+                    flavor=graalpy_version,
+                    implementation=self.implementation,
+                    filename=asset["name"],
+                    url=url,
+                )
+                # Only keep the latest GraalPy version of each arch/platform
+                if (python_version, arch, platform) not in results:
+                    results[(python_version, arch, platform)] = download
+
+        return list(results.values())
+
+    def _normalize_arch(self, arch: str) -> Arch:
+        return Arch(self.ARCH_MAPPING.get(arch, arch), None)
+
+    def _normalize_os(self, os: str) -> str:
+        return self.PLATFORM_MAPPING.get(os, os)
+
+    async def _fetch_checksums(self, downloads: list[PythonDownload], n: int) -> None:
+        checksums = {}
+        for batch in batched(downloads, n):
+            logging.info(
+                "Fetching GraalPy checksums: %d/%d", len(checksums), len(downloads)
+            )
+            checksum_requests = []
+            for download in batch:
+                url = download.url + ".sha256"
+                checksum_requests.append((url, self.client.get(url)))
+            for url, future in checksum_requests:
+                resp = await future
```

Comment:
These checksum requests are actually sequential, right?

Reply:
Not sure tbh 😅 I figured httpx is going to sent of the requests immediately so we get `n` requests in flight in parellel and then I handle them in batches of `n` before sending the next `n` requests.

Reply:
I think in this loop it's sending one request in one iteration, wait for it to finish, and then send the next one. So that’s sequential.

Instead, we could send requests in parallel using a TaskGroup, just like we did in the CPython example here: 
https://github.com/astral-sh/uv/blob/b33a19689c2d39f9bf2bccb37de91b33420906b3/crates/uv-python/fetch-download-metadata.py#L321-L331.

Reply:
I went with asyncio.gather because it's much shorter - let me know if this is ok